### ISSUE TITLE
ISSUE: when too many the spikes put into the neuron(the input is the same), the state is different according to the time(net.t)

### ISSUE THREAD

**Issue by zhouyanasd** [brian-team/brian2#943]
Here I found out a wired problem in Brian2, **The following is the result for this run, the input `data_train_s[0:1000]` is the same with `data_train_s[2000:3000]` and `data_train_s[1000:2000]` =0, but I got the different response from the network, the output is as following, the G.h at `1000*ms` is `[ 479.78008102.....` and at `3000*ms` is `[ 392.9473355...`:**
```
Starting simulation at t=0. s for a duration of 3.001 s
[ 0.  0.  0.  0.  0.]
[ 479.78008102  441.66822498  572.38234249  385.51723278  473.37427764]
[ 0.  0.  0.  0.  0.]
[ 392.9473355   362.45576778  469.8050491   316.11828296  387.76794325]
3.001 s (100%) simulated in < 1s
```

I only got this result when I set a large `S_in.w` and `S_e.w` which will make the number of  spike of in the network large. 
I guess the reason for this problem is the calculation error which related to the time 'net.t'.
I can not find exactly how this problem happened, could help to figure it out? Thank you very much and have a nice day!

The demo is as follows:
```python
######## this function is for input generation #############
import pandas as pd
class ST_classification_mass():
    def __init__(self, n_p, n, duration, frequency, dt):
        self.n_p = n_p
        self.n = n
        self.duration = duration
        self.frequency = frequency
        self.dt = dt
        self.n_1 = int(ceil(duration *dt * frequency))
        self.n_0 = int(ceil(duration)) - self.n_1
        self.D = int(duration / n)
        self.pattern_generation()

    def pattern_generation(self):
        self.pattern = []
        for i in range(self.n_p):
            b = [1] * self.n_1
            b.extend([0] * self.n_0)
            c = np.array(b)
            np.random.shuffle(c)
            self.pattern.append(c)

    def data_generation(self):
        value = []
        label = []
        for i in range(self.n):
            select = np.random.choice(self.n_p)
            fragment = np.random.choice(self.n)
            value.extend(self.pattern[select][self.D * fragment:self.D * (fragment + 1)])
            label.append(select)
            data = {'label': label, 'value': value}
        return data

    def data_noise_jitter(self, data):
        pass

    def data_generation_batch(self, number, noise = False):
        value = []
        label = []
        for i in range(number):
            data = self.data_generation()
            if noise :
                self.data_noise_jitter(data)
            value.append(data['value'])
            label.append(data['label'])
            df = pd.DataFrame({'value': pd.Series(value), 'label': pd.Series(label)})
        return df

    def get_series_data(self, data_frame, is_order=True, *args, **kwargs):
        if not is_order:
            data_frame_obj = data_frame.sample(frac=1).reset_index(drop=True)
        else:
            data_frame_obj = data_frame
        data_frame_s = []
        for value in data_frame_obj['value']:
            data_frame_s.extend(value)
        data_frame_s = np.asarray([data_frame_s]).T
        label = data_frame_obj['label']
        return data_frame_s, np.array(list(map(list, zip(*label))))
```
```python
######## this is network defination########
from brian2 import *
from brian2.synapses.synapses import SynapticPathway
    
start_scope()
prefs.codegen.target = "numpy"
defaultclock.dt = 1 * ms

ST = ST_classification_mass(2, 4, 1000, 100*Hz, 1*ms)
df_train = ST.data_generation_batch(10)
data_train_s, label_train = ST.get_series_data(df_train, False)
Time_array_train = TimedArray(data_train_s, dt=1*ms)

equ = '''
dv/dt = (I-v) / (30*ms) : 1 (unless refractory)
dh/dt = (-h)/(6*ms) : 1
I = (h)+13.5: 1
'''

neuron_in = '''
I = stimulus(t,i) : 1
'''

on_pre = '''
h+=w
'''
Input = NeuronGroup(1, neuron_in, threshold='I > 0', reset='I = 0', method='euler', refractory=0 * ms,
                    name = 'neurongroup_input')

G = NeuronGroup(10, equ, threshold='v > 15', reset='v = 13.5', method='euler', refractory=3 * ms,
                name='neurongroup_ex')
S_in = Synapses(Input, G, 'w : 1', on_pre=on_pre, method='euler', name='synapses_in')
S_e = Synapses(G, G, 'w : 1', on_pre=on_pre, method='euler', name='synapses_E')
S_in.connect(p=1)
S_e.connect(condition='i!=j', p=0.5)

G.v = '13.5'
G.h = '0'
S_in.w = 'rand()*120+120'
S_e.w = 'rand()*30+30'
# S_e.delay='randn()*3 * ms+3*ms'

@network_operation(dt=1000*ms)
def update_active():
    print(G.h[:5])
    for obj in net.objects:
        if isinstance(obj, Synapses) or isinstance(obj, NeuronGroup) or isinstance(obj, SynapticPathway):
            obj._restore_from_full_state(net._stored_state['third'][obj.name])

net = Network(collect())
net.store('init')
print('done')
```
```python
########## this is for run#########
data_train_s[1000:2000] = 0
data_train_s[2000:3000] = data_train_s[0:1000]
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.store('third')
net.run(3001*ms, report='text')
```

**Comment by thesamovar** [comment]
Difficult to say what's going on here. You could try monitoring Input.I 
and make sure that it is giving you exactly the same values as stimulus. 
It is possible for this to be off by one index over long time periods, 
but normally I'd expect it to be fine over a few seconds. If any single 
value is even slightly different, and if your network has sensitive 
dependence on initial conditions (likely given that it's recurrently 
connected) you'd expect to see a large divergence in the values printed, 
but by the looks of it you've tried to set everything up so that it 
should all be identical. Let us know the effect of monitoring Input.I 
and we'll see where we can go from there.

**Comment by zhouyanasd** [comment]
Thank you for your reply, The `Input.I` has been monitored many times in my formal simulation programme, the `Input.I` is guaranteed to be same for `data_train_s[0:1000]` and `data_train_s[2000:3000]`. The monitored results are shown as follows. From the results, we can find that 
```python
np.where(m_input.I[0][0:1000]==1),np.where(m_input.I[0][2000:3000]==1)
```
output:
```
((array([  6,  19,  23,  38,  40,  47,  56,  62,  64,  77,  79,  86, 102,
         104, 105, 108, 133, 149, 154, 162, 166, 170, 188, 190, 198, 199,
         208, 228, 230, 232, 233, 235, 236, 237, 246, 300, 307, 311, 319,
         343, 355, 356, 361, 368, 373, 374, 375, 381, 387, 399, 408, 413,
         418, 441, 459, 470, 478, 500, 501, 518, 533, 535, 555, 566, 574,
         581, 583, 600, 606, 629, 634, 640, 647, 655, 660, 668, 671, 675,
         678, 679, 685, 690, 698, 704, 713, 716, 726, 730, 737, 744, 784,
         786, 788, 797, 803, 811, 822, 842, 843, 852, 868, 883, 903, 909,
         931, 935, 942, 951, 980, 985, 989], dtype=int64),),
 (array([  6,  19,  23,  38,  40,  47,  56,  62,  64,  77,  79,  86, 102,
         104, 105, 108, 133, 149, 154, 162, 166, 170, 188, 190, 198, 199,
         208, 228, 230, 232, 233, 235, 236, 237, 246, 300, 307, 311, 319,
         343, 355, 356, 361, 368, 373, 374, 375, 381, 387, 399, 408, 413,
         418, 441, 459, 470, 478, 500, 501, 518, 533, 535, 555, 566, 574,
         581, 583, 600, 606, 629, 634, 640, 647, 655, 660, 668, 671, 675,
         678, 679, 685, 690, 698, 704, 713, 716, 726, 730, 737, 744, 784,
         786, 788, 797, 803, 811, 822, 842, 843, 852, 868, 883, 903, 909,
         931, 935, 942, 951, 980, 985, 989], dtype=int64),))
```

**Comment by mstimberg** [comment]
Hi. I just saw that your `network_operation` calls the private method `_restore_from_full_state` on some of the objects manually during a run -- this is definitely not something that we officially support. So if errors result from that specific usage, I'm afraid you are on your own...
That said, is there a reason why you use this instead of something along the lines of:
```Python
net.run(1000*ms)
net.restore('third')
net.run(1000*ms)
net.restore('third')
net.run(1000*ms)
...
```
?

If you can reproduce your problems with this standard usage of `net.store`/`net.restore` (and ideally with a simpler example), then we can look into it in more detail.

**Comment by zhouyanasd** [comment]
Yes, definitely, when I use `net.resotre`, the results for the two inputs are the same. For example, when I run:
```python
data_train_s[1000:2000] = 0
data_train_s[2000:3000] = data_train_s[0:1000]
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.store('third')
net.run(1001*ms, report='text')
net.restore('third')
net.run(1001*ms, report='text')
```

**However, the problem is still existing.**  If you run like this: 
```python
data_train_s[1000:2000] = 0
data_train_s[2000:3000] = 0
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.store('third')
net.run(3001*ms, report='text')
net.restore('third')
data_train_s[2000:3000] = data_train_s[0:1000] 
data_train_s[0:1000] = 0 ## before [2000:3000] we set the input is zero
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.run(3001*ms, report='text')
```
Before [2000:3000], I set the input to be zero, which means that there will nothing effect on the network before 2000*ms. And I got a different result: 
```
Starting simulation at t=0. s for a duration of 3.001 s
[ 0.  0.  0.  0.  0.]
[ 567.71667162  394.64865051  277.37610151  801.40794149  615.41707063]
[ 271.10251911  161.40935081   71.70460519  459.34438925  364.25231989]
[ 325.34266185  193.70291362   86.05072057  551.24654242  437.12917064]
3.001 s (100%) simulated in < 1s
Starting simulation at t=0. s for a duration of 3.001 s
[ 0.  0.  0.  0.  0.]
[ 0.  0.  0.  0.  0.]
[ 0.  0.  0.  0.  0.]
[ 496.47971832  353.65947934  261.99600214  677.77384062  515.44733786]
3.001 s (100%) simulated in < 1s
```
We can see that the same inputs led to different results in different time step(`[ 567.71667162  ....` and `[ 496.47971832 ...`). **Those results prove that the output(states) of the network is a function depends on the 'net.t'**, it is an confuse results for me to use Brian2 for computation, please help me to find what's going on here, and I need to make sure my simulation results are correct.

By the way, before run, the following codes will be delated from the program:
```python
    for obj in net.objects:
        if isinstance(obj, Synapses) or isinstance(obj, NeuronGroup) or isinstance(obj, SynapticPathway):
            obj._restore_from_full_state(net._stored_state['third'][obj.name])
```
Thank you very much, and best wishes.

**Comment by mstimberg** [comment]
Hi. I'm still having a hard time understanding the issue. Please try to strip down your example to the most simple (runnable) example that shows your issue. For example, I used the following which is based on your last input but making it runnable by adding random input and an input population:
```Python
from brian2 import *

data_train_s = np.zeros(3000)
data_train_s[:1000] = np.random.randn(1000)
data_train_s[1000:2000] = 0
data_train_s[2000:3000] = 0
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
group = NeuronGroup(1, "s = stimulus(t) : 1", threshold='s > 0')
mon = StateMonitor(group, 's', record=0, dt=1*ms)
spike_mon = SpikeMonitor(group)
net = Network(collect())
net.store('third')
net.run(3001*ms, report='text')
print(mon.s[0][:10])
print(spike_mon.t[:10])

net.restore('third')
data_train_s[2000:3000] = data_train_s[0:1000] 
data_train_s[0:1000] = 0 ## before [2000:3000] we set the input is zero
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.run(3001*ms, report='text')
print 'second run'
print(mon.s[0][2000:2010])
print(spike_mon.t[:10] - 2*second)
```

If I understood your issue correctly, then you seem to see that the results from 2s to 3s in the second run are not the same as the results for 0s to 1s in the first run, even though the input values are the same. With my simple example, I don't see this:
```console
Starting simulation at t=0. s for a duration of 3.001 s
3.001 s (100%) simulated in < 1s
[ 0.9717      1.40326362 -0.87337667  0.34442337 -1.21744854  0.17083509
 -1.80821845  2.00362928 -0.00909922 -0.82651516]
[ 0.   0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9] ms
Starting simulation at t=0. s for a duration of 3.001 s
3.001 s (100%) simulated in < 1s
second run
[ 0.9717      1.40326362 -0.87337667  0.34442337 -1.21744854  0.17083509
 -1.80821845  2.00362928 -0.00909922 -0.82651516]
[ 0.   0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9] ms
```

What additional elements do we need to reproduce the issue?

Note that your initially posted code does not run with Brian as it contains a syntax error:
```Python
Input = NeuronGroup(1, neuron_in, threshold='I > 0', reset='I = 0', method='euler', refractory=0 * ms,
                    name = 'neurongroup_input')
```
You cannot write to the variable `I` in the reset -- `I` is a subexpression that is calculated from the stimulus array, not something that you can assign to.

**Comment by zhouyanasd** [comment]
Hi, have a nice day and thank you for your reply. Firstly, what you mentioned about `reset ='I=0 '` is indeed not necessary for `Input`. I have removed it and then I found there is no effect on the result. 

Secondly, It is sorry for confusing you by a complicated example. To further explain the issue, I explored the example you offered: 

```python
from brian2 import *
start_scope()
prefs.codegen.target = "numpy"
# defaultclock.dt = 1 * ms
np.random.seed(100)

data_train_s = np.zeros(3000)
data_train_s[:1000] = np.random.randn(1000)
data_train_s[1000:2000] = 0
data_train_s[2000:3000] = 0
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train

equ = '''
dv/dt = (I-v) / (30*ms) : 1 (unless refractory)
dh/dt = (-h)/(6*ms) : 1
I = h: 1
'''

group = NeuronGroup(1, "s = stimulus(t) : 1", threshold='s > 0')
G = NeuronGroup(10, equ, threshold='v > 1.5', reset='v = 0', method='euler', refractory=3 * ms,
                name='neurongroup_ex')
S_in = Synapses(group, G, 'w = 2 : 1', on_pre="h+=w", method='euler', name='synapses_in')
S_e = Synapses(G, G, 'w = 2 : 1', on_pre="h+=w", method='euler', name='synapses_E')
S_in.connect(p=1)
S_e.connect(condition='i!=j', p=0.5)
mon = StateMonitor(group, 's', record=0, dt=1*ms)
mon_g = StateMonitor(G, 'v', record=0, dt=1*ms)
spike_mon = SpikeMonitor(group)
net = Network(collect())
net.store('third')
net.run(3001*ms, report='text')
print('mon.s is %s'%mon.s[0][500:510])
print('sipke_mon.s is %s' %spike_mon.t[0:10])
print('mon_g.v is %s' %mon_g.v[0][500:510])

net.restore('third')
data_train_s[2000:3000] = data_train_s[0:1000] 
data_train_s[0:1000] = 0 ## before [2000:3000] we set the input is zero
Time_array_train = TimedArray(data_train_s[0:3000], dt=1*ms)
stimulus = Time_array_train
net.run(3001*ms, report='text')
print('second run')
print('mon.s is %s'%mon.s[0][2500:2510])
print('sipke_mon.s is %s' %(spike_mon.t[0:10] - 2*second))
print('mon_g.v is %s' %mon_g.v[0][2500:2510])
```

Result: 
```
Starting simulation at t=0. s for a duration of 3.001 s
3.001 s (100%) simulated in 1s
mon.s is [ 0.87605119  2.15621999  1.28870256 -0.1425352   1.6659047  -1.3801143
  3.85793967  3.07557389  0.83081956 -0.87826817]
sipke_mon.s is [  1.   2.   4.   5.   6.   9.  11.  13.  14.  17.] ms
mon_g.v is [ 0.70952943  1.34381965  0.          0.          0.          0.80186207
  0.          0.          0.          0.82713256]
Starting simulation at t=0. s for a duration of 3.001 s
3.001 s (100%) simulated in < 1s
second run
mon.s is [ 0.87605119  2.15621999  1.28870256 -0.1425352   1.6659047  -1.3801143
  3.85793967  3.07557389  0.83081956 -0.87826817]
sipke_mon.s is [  1.   2.   4.   5.   6.   9.  11.  13.  14.  17.] ms
mon_g.v is [ 1.49203927  0.          0.          0.          0.93620704  0.          0.
  0.          0.73297022  0.        ]
```

A single neuron is not enough for this issue. In my opinion, the issue is something wrong when the spike transmit through synapses, especially with large number of spikes. Therefore, the larger `S_in.w` and `S_e.w`, the difference will be more obvious.

**Comment by mstimberg** [comment]
Hi. Thanks for the updated example, I can now finally reproduce your issue. After investigation I realized that it has nothing to do with the `TimedArray` or the synapses, but instead it is due to our implementation of the refractoriness mechanism. To determine whether a neuron is allowed to spike, we check whether the time that has passed since the last spike is greater than the refractory period. If the refractory period is an exact multiple of dt as in your case, this can sometimes go wrong (3ms after a spike the floating point representation of the passed time span might be 2.99999999ms or something like that) and the refractory period will therefore be one time step longer than expected. Due to the way floating point numbers work, whether this will happen or not depends on the absolute value of the time, explaining the difference between the two simulations that you saw.

I'll open a new issue for this to clearly state the problem. As a workaround until the fix is in place you could use e.g. `refractory=2.99*ms` instead of `refractory=3*ms` -- you shouldn't see any dependence on the absolute time anymore then.

**Comment by mstimberg** [comment]
An unrelated point:

> Firstly, what you mentioned about reset ='I=0 ' is indeed not necessary for Input. I have removed it and then I found there is no effect on the result, but it is a grate and worthful remind for me.

I'm surprised that you could run your code with the `I=0` in the first place -- didn't it raise an error about writing to a variable that you are not allowed to write to? Or maybe you are using an older release of Brian 2 where this was not yet the case?

**Comment by zhouyanasd** [comment]
Thank you for your reply and have a nice day. I fully understand how this issue happened, and the issue has been fixed by your advice, it is a great help for me, I am very appreciated for that. I will continue to focus on the new issue and learn more about how biran2 work.
The release I used is '2.0.2.1', and it didn't raise an error while the former programme running. Should it will raise an error in the recent release 2.1.2? If so, I think I should upgrade Brian2 to the 2.1.2 then.

**Comment by mstimberg** [comment]
> The release I used is '2.0.2.1', and it didn't raise an error while the former programme running. Should it will raise an error in the recent release 2.1.2? If so, I think I should upgrade Brian2 to the 2.1.2 then.

Yes, I think it should raise an error with the newer version. There have been various bug fixes and improvements since 2.0.2.1 (see http://brian2.readthedocs.io/en/stable/introduction/release_notes.html), so I'd recommend an update.

I will close this issue here as the problem will be dealt with in the new issue #949 -- thanks again for reporting it in the first place!
